### DISCUSSION TITLE
Add support for OpenTelemetry config

### DISCUSSION THREAD

**Discussion by matthewwjones** [renovatebot/renovate#28976] [Suggest an Idea]
### Background

We're using Renovate in a large enterprise.  It's great to see that there is support for capturing traces via OpenTelemetry - thank you for that!

We would like to be able to configure OpenTelemetry so that we can differentiate trace traffic originating from Renovate from the large volumes of trace traffic we have from other components.

### What I have tried

To do this, I've tried to do this using the following environment variables:
```
export OTEL_SERVICE_NAME="our-chosen-service-name"
export OTEL_RESOURCE_ATTRIBUTES="collection of comma separated key=value pairs"
```
These two settings are particularly important.

Checking the output in the logs after having enabled trace debugging with the `RENOVATE_TRACING_CONSOLE_EXPORTER` env var, I see:

```
  resource: {
    attributes: {
      'service.name': 'renovate',
      'telemetry.sdk.language': 'nodejs',
      'telemetry.sdk.name': 'opentelemetry',
      'telemetry.sdk.version': '1.24.0',
      'service.namespace': 'renovatebot.com',
      'service.version': '37.334.0'
    }
  },
```

The configuration provided in https://github.com/renovatebot/renovate/blob/main/lib/instrumentation/index.ts#L43 takes precendence over the `OTEL_*` env vars that we've set.

### What I'm requesting

I think exposing Telemetry configuration via the `config.js` would be great, or only setting the resource attributes above if `OTEL_RESOURCE_ATTRIBUTES` is not set in the environment.

**Comment by secustor** [comment]
To allow setting attributes using Admin/Bot configs which will take precedence to our variables sounds good. 

Don't like the OTEL_RESOURCE_ATTRIBUTES variant as that would not allow to selectively overwrite attributes e.g. the service name, but keep the version.

**Reply by matthewwjones** [reply]
Yep makes sense @secustor - your favoured approach would also keep all the config in a consistent place, too.

**Reply by secustor** [reply]
On second thought we do not have the variables at the point we are initializing the OTEL SDK.
So better to simply check if the env variable is already set and if this is not the case providing our defaults. 

For the three attributes that would be:
- `OTEL_SERVICE_NAME`
- `OTEL_SERVICE_NAMESPACE` ( not officially in specification, but used in e.g. in the Python SDK  https://pypi.org/project/opentelemetry_wrapper/) 
- `OTEL_SERVICE_VERSION` ( not implemented https://github.com/open-telemetry/opentelemetry-specification/issues/1901 )

**Reply by secustor** [reply]
Created an issue for it:

- https://github.com/renovatebot/renovate/issues/28986

**Reply by matthewwjones** [reply]
Thank you for writing up an issue @secustor. To check that I understand correctly - will your proposal allow us to set our own OTEL_RESOURCE_ATTRIBUTES too? 

This use case is pretty important for us as a mechanism to classify trace traffic.

**Reply by secustor** [reply]
The OTEL SDK should pick up that up already.

How I understood your problem is that it is not possible to overwrite the attributes above.m, correct?

**Reply by matthewwjones** [reply]
What I see at the moment is that `OTEL_RESOURCE_ATTRIBUTES` doesn't get picked up.  For example, I can do:

```shell
export OTEL_RESOURCE_ATTRIBUTES="foo=bar"
export RENOVATE_TRACING_CONSOLE_EXPORTER=true
```

and then:

`npx renovate`

But there is no sign of `foo=bar` in the verbose output:

```
{
  resource: {
    attributes: {
      'service.name': 'renovate',
      'telemetry.sdk.language': 'nodejs',
      'telemetry.sdk.name': 'opentelemetry',
      'telemetry.sdk.version': '1.24.0',
      'service.namespace': 'renovatebot.com',
      'service.version': '37.354.2'
    }
  },
  traceId: '3dd540a68d4a35eabc492c195a4fc6d6',
  parentId: undefined,
  traceState: undefined,
  name: 'run',
  id: '2b40f10e45d63d18',
  kind: 0,
  timestamp: 1715594415054000,
  duration: 12847512.9,
  attributes: {},
  status: { code: 0 },
  events: [],
  links: []
}
```

Does that help to clarify?

**Reply by secustor** [reply]
Yes, the reason is that we do not run the EnvDetector. That one is included in NodeAutoInstrumentations, which we currently not using. See https://github.com/renovatebot/renovate/issues/29026

**Reply by matthewwjones** [reply]
Understood, thank you for explaining and for raising the issue.
